lib/src/main/scala/spinal/lib/bus/amba4/axi/Axi4SharedOnChipRam.scala

```diff
@@ -60,60 +61,137 @@ case class Axi4SharedOnChipRam(dataWidth : Int, byteCount : BigInt, idWidth : In
-  }
+object Axi4SharedOnChipRamPort {
+    def apply(config: Axi4Config, ram: Mem[Bits]) = {
+        val port = new Axi4SharedOnChipRamPort(config)
+        port.attach(ram)
+        port
+    }
+
+    def apply(config: Axi4Config, ram: Mem[Bits], softReset: Bool) = {
+        val clock = ClockDomain.current.copy(softReset = softReset)
+        val domain = new ClockingArea(clock) {
+            val port = new Axi4SharedOnChipRamPort(config)
+            port.attach(ram)
+        }
+        domain.port
+    }
+}
 
-  val wordCount = byteCount / axiConfig.bytePerWord
-  val ram = Mem(axiConfig.dataType,wordCount.toInt)
-  val wordRange = log2Up(wordCount) + log2Up(axiConfig.bytePerWord)-1 downto log2Up(axiConfig.bytePerWord)
+case class Axi4SharedOnChipRamPort(config: Axi4Config) extends ImplicitArea[Axi4Shared] {
+    val axi                                = Axi4Shared(config)
+    override def implicitValue: Axi4Shared = this.axi
+
+    val ram = Handle[Mem[Bits]]
+
+    val logic = Handle {
+        new Area {
+            val wordRange       = config.wordRange
+            val arw             = axi.arw.unburstify
+            val writeDataStream = axi.writeData
+
+            val readStream      = Stream(Axi4R(axi.config))
+            val readAddrStream  = cloneOf(arw)
+            val writeStream     = Stream(Axi4B(axi.config))
+            val writeAddrStream = cloneOf(arw)
+
+            Vec(readAddrStream, writeAddrStream) <> StreamDemux(
+                arw,
+                U(arw.write),
+                2
+            )
+
+            val writeCombined = StreamJoin(writeAddrStream, writeDataStream)
+            val addr     = arw.addr(wordRange)
+            ram.write(
+                address = addr.resized,
+                data = writeCombined._2.data,
+                enable = writeCombined.fire,
+                mask = writeCombined._2.strb
+            )
+
+            writeStream.translateFrom(writeCombined) { (to, from) =>
+                if (axi.config.useId) to.id := from._1.id
+                if (axi.config.useWUser) to.user := from._1.user
+                to.setOKAY()
+            }
+            val writeReady = RegInit(False).riseWhen(True)
+            writeStream.ready := writeReady
+
+            val bStream      = cloneOf(writeStream)
+            val writeLast    = writeCombined.fire && writeCombined._1.last
+            val writePayload = RegNextWhen(writeStream.payload, writeLast)
+            val writeDone    = RegInit(False)
+            when(writeLast) {
+                writeDone := True
+            } elsewhen (bStream.fire) {
+                writeDone := False
+            }
+            bStream.valid := writeDone
+            bStream.payload := writePayload
+            axi.writeRsp << bStream
+
+            val readData = ram
+                .readSync(
+                    address = addr.resized
+                )
+                .resized
+            val savedReadData = Reg(cloneOf(readData))
```

Comment:
Why adding this buffer here ? Would it be possible to avoid it ?

Reply:
The key point is that when input stream is fired the address would changed, so that the data is updated one cycle after that. but the output might not fire right on the next cycle, so that the cache reg is necessary.